### ISSUE TITLE
User is authenticated but not connected

### ISSUE THREAD

**Issue by jmbldwn** [mscdex/node-imap#523]
I get this a lot, and have generally attributed it to some weirdness with office365 being flakey with IMAP, but the truth is I haven't had this problem with my previous implementation of my app, which used JAVAMAIL to connect to office365 via IMAP.

What's odd to me is that everything looks good until node-imap decides to send "NAMESPACE" to the server.  If the server was just trying to keep me from badgering it, I wouldn't expect to get this far.

Any ideas?

DEBUG OUTPUT:

DEBUG: [connection] Connected to host
DEBUG: <= '* OK The Microsoft Exchange IMAP4 service is ready. [QgBZADIAUABSADEAMQBDAEEAMAAwADcAMAAuAG4AYQBtAHAAcgBkADEAMQAuAHAAcgBvAGQALgBvAHUAdABsAG8AbwBrAC4AYwBvAG0A]'
DEBUG: => 'A0 CAPABILITY'
DEBUG: <= '* CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+'
DEBUG: <= 'A0 OK CAPABILITY completed.'
DEBUG: => 'A1 LOGIN "jim@domain.com" "<password>"'
DEBUG: <= 'A1 OK LOGIN completed.'
DEBUG: => 'A2 CAPABILITY'
DEBUG: <= '* CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CLIENTACCESSRULES CLIENTNETWORKPRESENCELOCATION CHILDREN IDLE NAMESPACE LITERAL+'
DEBUG: <= 'A2 OK CAPABILITY completed.'
DEBUG: => 'A3 NAMESPACE'
DEBUG: <= 'A3 BAD User is authenticated but not connected.'

**Comment by mscdex** [comment]
According to google, it seems to be a bug in Exchange's IMAP implementation where the server continues "happily" along if the username is valid but the password is not.

**Comment by jmbldwn** [comment]
But it does connect sometimes, so my password can't be incorrect.

**Comment by mscdex** [comment]
Do you have debug output from your other solution that shows how they are authenticating? Are they perhaps using `AUTHENTICATE PLAIN` instead of `LOGIN`? Is this a shared mailbox?

**Comment by jmbldwn** [comment]
Believe me, I would love to get debug output from the system that's running, but I've lost the formula to build and run it.  I never thought it would be hard to support a Scala/Play app long term. :-)

I have the code, though, and it looks like plain username/password auth, with TLS around it.  Same as what I'm doing now.  

It really is intermittent; feels like a timing issue at the root.  I'm trying a bunch of things now to shake it loose.

I have also verified my password multiple ways, and changed it just in case, so I don't think it's the bug we read about on the series of tubes.

It always comes after "NAMESPACE".  does that mean anything?

**Comment by mscdex** [comment]
What I meant is that there are really two ways to do simple username and password auth at the IMAP level: `LOGIN` (which you can see in this module's debug output) and `AUTHENTICATE PLAIN` (which is a SASL-based auth mechanism). `LOGIN` is pretty much supported everywhere and is pretty basic and straightforward. `AUTHENTICATE`-based mechanisms on the other hand could vary some implementation-wise. I had read that `AUTHENTICATE PLAIN` was actually broken on Office 365's IMAP server(s), but perhaps it's the reverse now (`LOGIN` is broken). Who knows.

If it's happening intermittently then I would definitely say it's a server-side issue, whether it's a straight up bug or some other issue in disguise (e.g. throttling of some kind).

The fact that it happens after trying to request the `NAMESPACE` information doesn't really matter as it would happen no matter what other useful requests you'd send (e.g. opening a mailbox, etc), at least according to other accounts of this problem I'm reading on google.

**Comment by jmbldwn** [comment]
I managed to revive my old Scala-based app enough to get some debug output from it.  I suspect now that it's running into the same issue, so I agree it's probably a server-side issue.

I think what's going on is Microsoft implemented a mechanism to shut down chatty clients.  They do not offer a definition of what is too chatty, so I don't yet know how to work around it, other than to retry until it works.  It doesn't look like they penalize you for trying; they just don't answer.  I would have preferred a message that says something like "give me a break and stop calling for a while" instead of the one I'm getting.

Part of the problem is I am connecting and disconnecting with each session, rather than holding sessions open.  In this application I am accessing a large number of accounts, so keeping a connection open seems like a bad idea, but maybe they prefer that over re-authenticating.

For background, here's the debug out from my previous Scala (javamail) app.

DEBUG: JavaMail version 1.5.1
DEBUG: successfully loaded resource: /META-INF/javamail.default.providers
DEBUG: Tables of loaded providers
DEBUG: Providers Listed By Class Name: {com.sun.mail.smtp.SMTPSSLTransport=javax.mail.Provider[TRANSPORT,smtps,com.sun.mail.smtp.SMTPSSLTransport,Oracle], com.sun.mail.smtp.SMTPTransport=javax.mail.Provider[TRANSPORT,smtp,com.sun.mail.smtp.SMTPTransport,Oracle], com.sun.mail.imap.IMAPSSLStore=javax.mail.Provider[STORE,imaps,com.sun.mail.imap.IMAPSSLStore,Oracle], com.sun.mail.pop3.POP3SSLStore=javax.mail.Provider[STORE,pop3s,com.sun.mail.pop3.POP3SSLStore,Oracle], com.sun.mail.imap.IMAPStore=javax.mail.Provider[STORE,imap,com.sun.mail.imap.IMAPStore,Oracle], com.sun.mail.pop3.POP3Store=javax.mail.Provider[STORE,pop3,com.sun.mail.pop3.POP3Store,Oracle]}
DEBUG: Providers Listed By Protocol: {imaps=javax.mail.Provider[STORE,imaps,com.sun.mail.imap.IMAPSSLStore,Oracle], imap=javax.mail.Provider[STORE,imap,com.sun.mail.imap.IMAPStore,Oracle], smtps=javax.mail.Provider[TRANSPORT,smtps,com.sun.mail.smtp.SMTPSSLTransport,Oracle], pop3=javax.mail.Provider[STORE,pop3,com.sun.mail.pop3.POP3Store,Oracle], pop3s=javax.mail.Provider[STORE,pop3s,com.sun.mail.pop3.POP3SSLStore,Oracle], smtp=javax.mail.Provider[TRANSPORT,smtp,com.sun.mail.smtp.SMTPTransport,Oracle]}
DEBUG: successfully loaded resource: /META-INF/javamail.default.address.map
getting store
DEBUG: getProvider() returning javax.mail.Provider[STORE,imaps,com.sun.mail.imap.IMAPSSLStore,Oracle]
DEBUG IMAPS: mail.imap.fetchsize: 16384
DEBUG IMAPS: mail.imap.ignorebodystructuresize: false
DEBUG IMAPS: mail.imap.statuscachetimeout: 1000
DEBUG IMAPS: mail.imap.appendbuffersize: -1
DEBUG IMAPS: mail.imap.minidletime: 10
DEBUG IMAPS: enable STARTTLS
connecting to store
DEBUG IMAPS: protocolConnect returning false, host=outlook.office365.com, user=jim@domain.com, password=<null>
creating PasswordAuthentication object
DEBUG IMAPS: trying to connect to host "outlook.office365.com", port 993, isSSL true
* OK The Microsoft Exchange IMAP4 service is ready. [QgBZADIAUABSADEAMQBDAEEAMAAwADYANgAuAG4AYQBtAHAAcgBkADEAMQAuAHAAcgBvAGQALgBvAHUAdABsAG8AbwBrAC4AYwBvAG0A]
A0 CAPABILITY
* CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
A0 OK CAPABILITY completed.
DEBUG IMAPS: AUTH: PLAIN
DEBUG IMAPS: AUTH: XOAUTH2
DEBUG IMAPS: protocolConnect login, host=outlook.office365.com, user=jim@baldw.in, password=<non-null>
DEBUG IMAPS: AUTHENTICATE PLAIN command trace suppressed
DEBUG IMAPS: AUTHENTICATE PLAIN command result: A1 OK AUTHENTICATE completed.
A2 CAPABILITY
* CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CLIENTACCESSRULES CLIENTNETWORKPRESENCELOCATION BACKENDAUTHENTICATE CHILDREN IDLE NAMESPACE LITERAL+
A2 OK CAPABILITY completed.
DEBUG IMAPS: AUTH: PLAIN
DEBUG IMAPS: AUTH: XOAUTH2
getting server folders
A3 LIST "" "%"

**Comment by mscdex** [comment]
Ah yes, that is using `AUTHENTICATE PLAIN` instead of `LOGIN`.  Does the `LIST` request shown in the debug output actually complete successfully or does the server respond with an error?

**Comment by jmbldwn** [comment]
It completes successfully.

The old app does fail occasionally, but i have a different retry period on that app so the problem wasn't so significant.  When Microsoft is ignoring me, it gives me the same "authenticated but not connected" error.

**Comment by jmbldwn** [comment]
I'm suspicious that the core problem could be how I'm leaving the connections.  Admittedly I haven't worried about whether I am closing things down yet.  Is there a best practice on closing connections?

**Comment by mscdex** [comment]
You should just have to call `imap.end()` to close the connection.

**Comment by jmbldwn** [comment]
Events continue to dribble in after that.  After I call map.end() I get this:

DEBUG: => 'A44 LOGOUT'
DEBUG: <= '* BYE Microsoft Exchange Server 2016 IMAP4 server signing off.'
DEBUG: <= 'A44 OK LOGOUT completed.'
DEBUG: [connection] Error: Error: read ECONNRESET
DEBUG: [connection] Closed

I think these messages are ignorable, but it also seems like we could have expected the connection to be reset.  It may be my flavor of OCD, but I get uncomfortable ignoring errors.

Is there a way to do this more cleanly?

**Comment by mscdex** [comment]
Node way back when used to swallow some ECONNRESETs, but I've seen this reported before and it seems to only happen to particular servers. AFAIK there's not much else this module can do about those since it is properly logging out as you can see. If you can, do a packet capture (at least of the logout) and verify that the server is sending the RST.

**Comment by jmbldwn** [comment]
I'm closing this because I've been able to narrow it down to a confusing way to limit connections by office365.  I increased the minimum time between connections and keep my connections open longer and I'm no longer experiencing this.

So, if you were thinking like I was that you could open a connection to office365, do some stuff, and close it, and repeat that frequently, you might want to think about structuring your code differently.

**Comment by mscdex** [comment]
Thanks for the update and helpful tips!

**Comment by jmbldwn** [comment]
Here's my more detailed suspicion:

Office365 says the limit on concurrent IMAP connections is 20.  I wasn't trying to do more that one simultaneous connection, so it shouldn't have been giving me any trouble.  Indeed it works fine at first if I haven't connected in a while, but after some number of connections (I was never able to accurately measure this) it starts exhibiting this problem.

The leading explanation is that somehow when I close the connection it isn't getting closed down on the server side, so my number of concurrent connections racks up to the limit until the server gets around to cleaning up unused connections.  It may just be a simple accounting bug on the server.

I realized that IMAP might not be meant for this kind of "guerrilla" use case, so I changed my system to keep a single connection open indefinitely instead of closing and re-opening, and this problem has disappeared.
